### PR TITLE
wrap external api calls and db actions with service wrapper, refs #254

### LINKED ISSUE #254: write mock providers for all required services
as a developer, in order to work on faucet, i want to be able to test code that relies on network services.

as a QA tester, in order to make sure that faucet works, i want to be able to run faucet in a mode that uses mock services rather than the real ones so i can try out the whole signup process.

to implement, we can use a service dependency injection architecture or something similar.

services:

- [x] twilio
- [x] sendgrid
- [x] recaptcha 
- [x] ip rate limiting
- [x] steem blockchain
- [x] mysql db

### PR DESCRIPTION
**Pull request by gl2748**
Closes #254.

### CODE REVIEW

**Comment by gl2748** [comment]
Email activation Code: http://localhost:3000/confirm-email?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ0eXBlIjoiY29uZmlybV9lbWFpbCIsImVtYWlsIjoiZm9vQGJhci5jb20iLCJpYXQiOjE1MjIwNzk4ODUsImV4cCI6MTYwODM5MzQ4NX0.OMHsvMQSiZlHSZCEoUbk3QvaSAZUbgo0-1-k0HmXRSU
Use phone code: 123

**Comment by gl2748** [comment]
An additional request for this work was that in production a certain set of credentials, username, email, phone, are excepted from the normal validation and end in an off-chain non-event 'demo signup procedure complete :-)'

**Comment by gl2748** [comment]
@cyon1c ok, I have now done that.

**Comment by cyon1c** [comment]
Tested locally, appears to be working great!

File: routes/api.js [hunk 1]
```diff
@@ -13,13 +13,16 @@ const badDomains = require('../bad-domains');
...
 const db = require('./../db/models');
+const utils = require('../helpers/utils');
 
 const { Sequelize } = db;
 const { Op } = Sequelize;
 
 const conveyorAccount = process.env.CONVEYOR_USERNAME;
 const conveyorKey = process.env.CONVEYOR_POSTING_WIF;
 
+const env = process.env.DATABASE_NAME;
```
Comment (jnordberg) [inline routes/api.js]: Let's call this something more obvious, like `DEBUG_MODE` or something. And do a logger.warn("IN DEBUG MODE") so we know when it is active

File: helpers/utils.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+                'conveyor.set_user_data': true,
+            };
+            return endpoints[endpoint];
+        },
+    };
+    const fnToCall = env === 'development' ? services[service] : fn;
+    return fnToCall;
+};
+
+const databaseWrapper = (action, fn, env) => {
```
Comment (jnordberg) [inline helpers/utils.js]: Ideally we should test the database logic in debug mode as well

File: routes/api.js (hunk 1 above)
Comment (gl2748) [reply routes/api.js]: Ok.

File: helpers/utils.js (hunk 2 above)
Comment (gl2748) [reply helpers/utils.js]: Ok. Will do.

**Comment by gl2748** [comment]
All links to get to next steps are logged to server console like this:

screenshot from 2018-04-04 17-25-52

**Comment by gl2748** [comment]
@jnordberg thoughts on this one?

**Comment by jnordberg** [comment]
Looks like it does what we want, I would mock the services the other way around though. Instead of having a service wrapper where the caller has to worry about passing the correct debug state I would write abstractions for the actions we want to perform instead.

So instead of

```js
const emailRegistered = await utils.serviceWrapper('steem', steem.api.signedCallAsync, DEBUG_MODE)(
    'conveyor.is_email_registered', [req.body.email],
    conveyorAccount, conveyorKey,
);
```

the handler can do

```js
const emailRegistered = await checkEmailRegistered(req.body.email)
```

and the `checkEmailRegistered` implementation would handle mocking in debug mode. This approach allows handlers not to worry about passing correct state possibly ending up in 'debug mode' when we don't want it and also allows for easier refactoring in the future if we change providers for e.g. sms

File: routes/api.js [hunk 3]
```diff
@@ -344,12 +267,12 @@ router.post('/request_sms', apiMiddleware(async req => {
...
     if (
         user.last_attempt_verify_phone_number &&
-    user.last_attempt_verify_phone_number.getTime() > Date.now() - (2 * 60 * 1000)
+        user.last_attempt_verify_phone_number.getTime() > Date.now() - (2 * 60 * 1000)
     ) {
         throw new ApiError({ field: 'phoneNumber', type: 'error_api_wait' });
     }
 
-    const phoneExists = await req.db.users.count({
+    const phoneExists = db.users.count({
```
Comment (bnchdrff) [inline routes/api.js]: we should `await` the return of `count` right?

Comment (jnordberg) [reply routes/api.js]: yes, good catch

File: helpers/services.js [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+const conveyorKey = getEnv('CONVEYOR_POSTING_WIF')
+const createAccountDelegation = getEnv('CREATE_ACCOUNT_DELEGATION')
+const createAccountDelegator = getEnv('DELEGATOR_USERNAME')
+const createAccountFee = getEnv('CREATE_ACCOUNT_FEE')
+const createAccountWif = getEnv('DELEGATOR_ACTIVE_WIF')
+const recaptchaSecret = getEnv('RECAPTCHA_SECRET')
+
+const rpcNode = getEnv('STEEMJS_URL')
+if (rpcNode) {
+    steem.api.setOptions({ url: rpcNode })
```
Comment (cyon1c) [inline helpers/services.js]: This is failing for me - `steem` is undefined in debug-mode here.

**Pull request merged**